java-checks/src/main/java/org/sonar/java/checks/security/StandardInputReadCheck.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2018 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.security;
+
+import com.google.common.collect.ImmutableList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Predicate;
+import org.sonar.check.Rule;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.java.matcher.NameCriteria;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodReferenceTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S4829")
+public class StandardInputReadCheck extends IssuableSubscriptionVisitor {
+
+  private static final List<MethodMatcher> METHOD_MATCHERS = Arrays.asList(
+    MethodMatcher.create().typeDefinition("java.lang.System").name("setIn").withAnyParameters(),
+    MethodMatcher.create().typeDefinition("java.io.Console").name(NameCriteria.startsWith("read")).withAnyParameters());
+
+  private static final MethodMatcher CLOSE_METHOD = MethodMatcher.create().name("close").withoutParameter();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.METHOD_INVOCATION, Tree.Kind.METHOD_REFERENCE, Tree.Kind.MEMBER_SELECT);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+
+    if (tree.is(Tree.Kind.METHOD_INVOCATION)) {
+      checkMethod(tree, methodMatcher -> methodMatcher.matches((MethodInvocationTree) tree));
+    } else if (tree.is(Tree.Kind.METHOD_REFERENCE)) {
+      checkMethod(tree, methodMatcher -> methodMatcher.matches((MethodReferenceTree) tree));
+    } else {
+      checkMemberSelect((MemberSelectExpressionTree) tree);
```

Comment:
Checking member select is not enough, we should focus on identifier. Its probably also possible to only write the test on subscribing to identifiers

```
import static java.lang.System.in; // this will probably raise an issue

class A {
  void foo() throws IOException {
    in.read(); // but this should raise an issue as well, it's an usage of "in"
  }
}
```

Reply:
Indeed, focusing on identifiers will enable us to catch all cases